ipcore/vhdl_node.py

```diff
@@ -0,0 +1,373 @@
+        self.signal = signal
+    def generate(self):
+        """Generate a VHDL port."""
+        if self.signal.length == 1:
+            type_declaration = self.signal.data_type
+        else:
+            type_declaration = f"{self.signal.data_type}({(self.signal.length - 1)} downto 0)"
+        return f"{self.name.ljust(32)} : {self.direction.value} {type_declaration}"
+    @property
+    def length(self):
+        return self.signal.length
+    @property
+    def data_type(self):
+        return self.signal.data_type
+    @property
+    def underlying_data_type(self):
+        return self.signal.underlying_data_type
+    def generate_assignment(self, new_signal_value):
+        """Generate VHDL assignment to this port, including necessary resizing.
+
+        Resizes using underlying data type if possible.
+        VHDL data types must match except for std_logic and std_logic_vector.
+        For assignmnents between std_logic and std_logic_vector,
+        the first bit is where data is assumed to be stored.
+
+        Parameters
+        ----------
+        new_signal_value : Signal
+            signal being assigned to this. VHDL data types must match except for std_logic and std_logic_vector
+
+        Returns
+        -------
+        str
+            VHDL code to assign new_signal_value to this port
+        """
+        return self.signal.generate_assignment(new_signal_value)
+
+class PortDir(Enum):
+    In = "in "
+    Out = "out"
+
+class ArraySignal(Signal):
+    def __init__(self, name, vhdl_type, length, element_type, element_length, element_data_type=None):
+        super().__init__(name, length, None, vhdl_type)
+        self.signals = [Signal(f"{self.name}({i})", element_length, None, element_type, element_data_type) for i in range(length)]
+    def __getitem__(self, key):
+        return self.signals[key]
+class Entity(BaseVHDLNode):
+    def __init__(self, name, ports):
+        super().__init__(name)
+        self.ports = ports
+    def generate(self):
+        """Generate a VHDL entity."""
+        vhdl_string = f"entity {self.name} is \n"
+        vhdl_string += self._generate_ports()
+        vhdl_string += f"end entity {self.name};\n\n"
+        return vhdl_string
+    def _generate_ports(self):
+        vhdl_string = ""
+        vhdl_string += tab() + "port (\n"
+        vhdl_string += tab(2)
+        ports_vhdl = [port.generate() for port in self.ports]
+        vhdl_string += (";\n" + tab(2)).join(ports_vhdl)
+        vhdl_string += "\n"
+        vhdl_string += tab() + ");\n"
+        return vhdl_string
+    def get_port(self, name):
+        """Get entity port by name.
+
+        Parameters
+        ----------
+        name : str
+            Name of port to retrieve
+
+        Returns
+        -------
+        Port, None
+            Returns entity port with the given name if found, otherwise None.
+        """
+        return next((port for port in self.ports if port.name == name), None)
```

Comment:
I'm a bit confused by using `next` here. Are we expecting there to be multiple ports where `port.name == name` is true? If there is only one port that matches, then is there any need to use an iterator? 

That said, I've never actually used `next()` before, so I could be missing something.

Reply:
Next returns the next item that the condition is true for. Given I never reuse iterators, it is equivalent to finding the first item that it is true for and returning it.

>  then is there any need to use an iterator

How else would you get the port by name?

Reply:
Okay, that makes sense. You make build a generator that contains all of the ports where `port.name == name`, then return the first one with `next`. I am not very familiar with generators, but that seems like the idiomatic thing to do.